app/src/main/java/com/lagradost/quicknovel/ui/result/ResultViewModel.kt

```diff
@@ -77,9 +91,289 @@ class ResultViewModel : ViewModel() {
             )
         }
 
+        // Invalidate cache for instant filter updates
+        invalidateStatusCache()
+        return true
+    }
+
+    fun markAllPreviousChapters(chapter: ChapterData, value: Boolean): Boolean {
+        val streamResponse = (load as? StreamResponse) ?: return false
+        val currentIndex = chapterIndex(chapter) ?: return false
+        
+        // Mark all chapters from index 0 to currentIndex (inclusive) as read/unread
+        for (i in 0..currentIndex) {
+            if (value) {
+                setKey(
```

Comment:
While this is idiomatic, using setKey and removeKey is actually quite an expensive operations due to calling apply between each call. While I do not expect you to fix this, do note that DataStore.editor should be used in this case.